Move caret over closing paren

hellyab opened the pull request:
Move caret over the closing paren when ")" is inputted right before to the closing paren
closes #131

hellyab committed:
[build] move carret out of parrens for TypAnnZA

hellyab committed:
[build] add support for nested parens in TypeAnnZA

hellyab committed:
[fix] construct the correct ZTyp instead of using move_right

hellyab committed:
[build] add closing parens logic to Action_Pat in ana

hellyab committed:
[build] add close parens rules

hellyab committed:
[build] add close braces functionality for lambdas

hellyab committed:
[build] fix closing braces for lambda

hellyab committed:
[build] add closing parens and braces to the undo history

hellyab committed:
[cleanup] reorder rules

cyrus- commented on src/hazelcore/Action_Exp.re:
```diff
@@ -1912,9 +1979,8 @@ and syn_perform_operand =
-  | (Init, _) => failwith("Init action should not be performed.")
-  };
-}
+  | (Construct(SCloseBraces), _) => Failed
```
why does SCloseBraces but not SCloseParens have a fallthrough case like this?

hellyab replied:
The one for SCloseParens was explicitly put as (Construct(SCloseParens), CursorE(_)) => Failed, doing the same for this one too

cyrus- commented on src/hazelcore/Action_Exp.re:
```diff
@@ -3233,6 +3361,8 @@ and ana_perform_operand =
       }
     }
 
+  | (Construct(SCloseBraces), _) => Failed
```
same as above comment

hellyab replied:
> Looks pretty good, just a few small things to respond to / revise. Also, can you add support for `]` when in a list type like `[Int]`? (We'll have list literals in a future pull too, so this can set up for that as well.)

Done!

cyrus- commented on src/hazelcore/ZTyp.re:
```diff
@@ -168,7 +168,8 @@ let move_cursor_right_zoperator: zoperator => option(zoperator) =
   zty |> move_cursor_right_zopseq
-and move_cursor_right_zopseq = zopseq =>
+and move_cursor_right_zopseq = zopseq => {
+  print_endline("ZTyp move_cursor_right_zopseq");
```
don't leave debug prints when you mark as ready for review

hellyab replied:
I thought I removed these. My bad.

cyrus- commented on src/hazelweb/model/UndoHistory.re:
```diff
@@ -464,7 +464,13 @@ let get_new_action_group =
+    * it is an edit action, therefore there is a need to make sure that
+    * the action is not Construct(SCloseParens) or Construct(SCloseBraces)
+   */
+  if (UndoHistoryCore.is_move_action(new_cursor_term_info)
```
can we make this a predicate on the action instead of having it compare the previous and new states?

hellyab replied:
The entire if expression looks like it was redundant because the else branch was handling all possible cases. I just removed the if branch. Not sure why it was added, but I have removed it at the moment. It was added in dab69adceb8634f31e334ef922287795a66c0c03

cyrus- reviewed (changes requested):
Looks pretty good, just a few small things to respond to / revise. Also, can you add support for `]` when in a list type like `[Int]`? (We'll have list literals in a future pull too, so this can set up for that as well.)

hellyab committed:
[cleanup] put explicit cases for SCloseBrace
[cleanup] remove print_endline

hellyab committed:
[build] add support for closing square brackets on list type

hellyab committed:
[cleanup] make undo history predicate based on action

hellyab commented:
> Looks pretty good, just a few small things to respond to / revise. Also, can you add support for `]` when in a list type like `[Int]`? (We'll have list literals in a future pull too, so this can set up for that as well.)

Thanks! I have addressed the issues and also added support for `]`. It is working fine on the list type.

hellyab committed:
[cleanup] remove unnecessary changes and put comments in their right place

cyrus- commented on src/hazelweb/model/UndoHistory.re:
```diff
@@ -464,195 +464,195 @@ let get_new_action_group =
+      | Some(zrules_before) =>
         switch (
-          CursorInfo_Exp.get_outer_zrules(new_cursor_term_info.zexp_before)
+          CursorInfo_Exp.get_outer_zrules(new_cursor_term_info.zexp_after)
```
@hellyab was there a bug here (unrelated to the PR)?

hellyab replied:
@cyrus- it has always been `zexp_after`, I think git is comparing it to another line.

cyrus- committed:
Merge branch 'dev' into feat-close-parens

cyrus- committed:
merge with dev
